aten/src/ATen/native/ForeachUtils.h

```diff
@@ -13,7 +13,7 @@ void check_foreach_api_restrictions(TensorList tensors) {
   }
 }
 
-void check_foreach_api_restrictions(TensorList tensors, ArrayRef<double> scalars) {
+void check_foreach_api_restrictions(TensorList tensors, ArrayRef<Scalar> scalars) {
```

Comment:
I'm slightly confused about what the difference is between `check_foreach_api_restrictions` and `can_use_fast_path`. It would be nice to have some comments in the code that delineate what each is responsible for.

From my understanding:

check_foreach_api_restrictions
- Tensor lists must be non-empty.
- All tensors in all lists must have the same dtype.
- All TensorLists and ScalarLists must have the same number of elements.
- Corresponding tensors must have the same size.

`can_use_fast_route`
- all tensors must have the same device
- all tensors must be non-overlapping and strided
- corresponding tensors must have the same strides
- type promotion must not occur (and the only time this can occur is TensorList-ScalarList operations)

Is that correct? am I missing anything?

Reply:
`check_foreach_api_restrictions` checks API inputs. If restrictions are not met - error is thrown. 
`can_use_fast_path` checks if we can go via 'fast' path. if not, go via slow path. no errors. 

You are correct and i dont think you are missing anything. Im planning to add an additional PR with code docs only but once next few PRs will land. The reason for that is that this logic will change a lot. 

--
Added a small comment around `check_foreach_api_restrictions`. `can_use_fast_path` already has one.